### PR TITLE
EZP-23556: Named queries specification

### PR DESCRIPTION
**Pull request by bdunogier**
> Story: https://jira.ez.no/browse/EZP-24456
> Status: draft / work-in-progress

Comments welcome.

### CODE REVIEW

File: doc/specifications/proposed/named_queries/named_queries.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+All REST calls on `/content/views`.
+
+### Query controller
+```twig
+{{ render( controller( ez_query, {'identifier': 'latest_articles', 'parameters': {'type': 'article'}} ) ) }}
+```
+
+### Services
+
+A main `QueryService` / `ezplatform.query_service`, with methods to load/create/update/delete named queries.
```
Comment (andrerom) [inline doc/specifications/proposed/named_queries/named_queries.md]: Isn't this mixing named queries and persisted queries?
They are kind of different.

File: doc/specifications/proposed/named_queries/named_queries.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+## Overview
+
+Named queries are Content or Location Query objects that are predefined in the system.
+
+They can be used in multiple ways:
+- from twig templates, using a sub-request: {{ render( controller( ez_query, {'identifier': 'latest_articles' ) ) }}
+- from php code, by getting the Query from a service
+- from REST, by executing a predefined view / query
+
+They
```
Comment (crevillo) [inline doc/specifications/proposed/named_queries/named_queries.md]: unneded it seems?

File: doc/specifications/proposed/named_queries/named_queries.md (hunk 1 above)
Comment (lolautruche) [inline doc/specifications/proposed/named_queries/named_queries.md]: AFAIR of our discussions, they're no different in essence. 
One is defined as a *type* (like a FormType), the other is persisted. In both cases, the result of the executed query is a *view*.

File: doc/specifications/proposed/named_queries/named_queries.md (hunk 2 above)
Comment (bdunogier) [inline doc/specifications/proposed/named_queries/named_queries.md]: "They" probably do something, but it's very vague right now, you're right :-)

File: doc/specifications/proposed/named_queries/named_queries.md (hunk 1 above)
Comment (bdunogier) [inline doc/specifications/proposed/named_queries/named_queries.md]: If I remember the conversation we had with @tobyS a couple weeks back, we had concluded that named and persisted queries were in essence the same thing, indeed. 

Named (php classes based) queries would be read-only (we mentioned throwing an  HTTP error when such an operation is tried). Persisted queries could "inherit" from named queries. We also discussed priorities/overriding an existing named query.

Comment (pspanja) [inline doc/specifications/proposed/named_queries/named_queries.md]: With persisted queries, why do we need named (class based) queries at all? To me it seems like something that we can leave to user implementation.

Comment (bdunogier) [inline doc/specifications/proposed/named_queries/named_queries.md]: We don't have persisted queries yet, and they're not a high priority. The idea behind named queries is to standardize "content listing" in general.

Yes, queries can be added as flat files already, and it is an implementation detail. What we add on top of this is:
- Standardization of those query files (I'm considering auto-registration based on the query's path, like Sf2 controllers)
- Template of query results (just pushed a commit that introduces this)

The goal is to fix, at the same time, the "list" controller people have been missing for a while.

Comment (pspanja) [inline doc/specifications/proposed/named_queries/named_queries.md]: Ok, but then do you agree we don't let that (named queries) knowledge into PAPI and it's REST endpoint?

Comment (bdunogier) [inline doc/specifications/proposed/named_queries/named_queries.md]: Errr no. Why would we not allow them ?

Comment (pspanja) [inline doc/specifications/proposed/named_queries/named_queries.md]: Just to make sure we are on the same line, allowing them would mean one could use `QueryService::load( $identifier )` to load named and persisted queries, and there would also be some kind of override system between the two?

Then that would also mean REST resource `/content/views/<identifier>/results` would behave in analogous way...

Comment (bdunogier) [inline doc/specifications/proposed/named_queries/named_queries.md]: I'm gonna try to move this conversation to the PR itself, as this commit is outdated.

**Comment by bdunogier** [comment]
Just to make sure we are on the same line, allowing them would mean one could use QueryService::load( $identifier ) to load named and persisted queries, and there would also be some kind of override system between the two?

Then that would also mean REST resource /content/views/<identifier>/results would behave in analogous way...

**Comment by andrerom** [comment]
Well it's the same point I have had as well several times when we have discussed this.

It looks inconsistent / wrong to handle both cases with same API + SPI, named query is a extension point while persisted queries are, well, persistence and will need spi impl. And as @pspanja hinted about might be represented in a different way then the query class (id + machine readable query string for instance).

**Comment by pspanja** [comment]
> Named query is a extension point in MVC stack imo, while persisted queries are, well, a Repository concern, and will need api+spi impl

That is my POV exactly as well.

**Comment by bdunogier** [comment]
Okay, let's split them, then.

I guess we won't expose those over REST, right ? They were meant for persisted queries, and don't apply. If we want to expose it, we need to design this new API.

We also don't need to put this into the Public API, as it's completely separated from the Repository. What about QueryHelper for the name ?

**Comment by andrerom** [comment]
Before we just change it, could you first re iterate the arguments for keeping them handled in the same way? As you have been discussing this in the past I'm sure there are concerns me and @pspanja overlooks here.

> I guess we won't expose those over REST, right ?

For Platform UI we kind of need to expose it somehow, I'm not sure how actually. It is a bit different then the normal view given it won't support letting user build query and post it => store it, retrieve, update, delete. But it should still have most of the other features of a "View", even if that is not super well defined or implemented yet (possibility to represent output in other ways, cache handling, ..).

**Comment by pspanja** [comment]
What I would like to see here is why we need an implementation parallel to persisted queries in the first place. For what reasons this use-case can't be handled by persisted queries?

**Comment by andrerom** [comment]
> What I would like to see here is why we need an implementation parallel to persisted queries in the first place. For what reasons this use-case can't be handled by persisted queries?

Good point, it should be in the document. Afaik main reasons is to allow queries that are not editorial, hence rather written by developers for use in UI and in website. Typical use case is children view, and several others.

**Comment by bdunogier** [comment]
> For what reasons this use-case can't be handled by persisted queries?

The main reason would be that we don't have them yet, and they're not even planned.

As André explained, the target is project developers, not content editors. The principle is the same than fetch aliases that existed in legacy, and allowed you to create custom fetch content queries for a project that could be reused.

**Comment by pspanja** [comment]
> Do you agree with the principle ? It might be that this is premature work, and that we should wait for persisted queries.

I agree. AFAICS the question here is query *parameterization*, for this to work with persisted queries we should somehow support it. I'm not sure we want that in the API, but do you have an idea how it could work, on top of it maybe?

Otherwise, I think big part of this is providing a good developer experience. In my POV it would be good if it could be done in MVC on top of persisted queries, but it might be too limiting or require pushing more functionality to API, which we should try to avoid.

**Comment by bdunogier** [comment]
> Otherwise, I think big part of this is providing a good developer experience.

That's actually the main point here, as it is one of our business focuses.

> From my POV it would be good if it could be done in MVC on top of persisted queries.

We do not have Persisted Queries, and we don't have plans to add them anytime soon, as they require SPI changes.

Why would it be better if it only relied on persisted queries ? What is the problem exactly with queries defined in files ?

> but it might be too limiting or require pushing more functionality to API, which we should try to avoid.

Well, I have to disagree here. We have new versions for a reason, and we have changed the API when we needed to.
We have a REST API feature (views) that requires this. Unless we want to use the MVC layer within the REST API, we need to add something to the API itself.

**Comment by pspanja** [comment]
> Why would it be better if it only relied on persisted queries ?

First because we would not have parallel implementations.

> What is the problem exactly with queries defined in files ?

Well depends on implementation. I would not mix it with the persisted queries, as the storage would be different, also we could end up having API for fetching stuff from MVC configuration.

> Well, I have to disagree here. We have new versions for a reason, and we have changed the API when we needed to.

I meant that as a general rule, we definitely should not be hasty about adding stuff to API. But if there happens to be a good use case, I won't be against it.

**Comment by lolautruche** [comment]
Correct me if I'm wrong @andrerom , but as repeated several times, **persisted queries are not specified, prioritized or planned**. We're not talking about that here. 
Why sticking on this? If we have to wait that everything we have in mind to be implemented, we'll be releasing ezplatform 1.0 in 2 or 3 years!

**Comment by bdunogier** [comment]
I'm sorry to say that the goal isn't to avoid having parallel implementations.

In terms of developer experience, I think that being able to store a query in a php class, and use it from my code, is better than having to create it in the database somehow. Unless we provide a great user experience for this, it will be tedious.

Again, parallel implementations are not a problem for the developer, they're a problem for us. We are not writing software to assist ourselves, but to help our users. In that case, the developers.

We already have leaks from MVC to the API, mostly in the REST API.

**Pull request closed without merging**